tiledb/sm/subarray/range_subset.h

```diff
@@ -0,0 +1,328 @@
+   refactor).
+   * @param new_range The range to add.
+   */
+  virtual Status add_range_unsafe(const Range& range) = 0;
+
+  virtual const Range& get_range(const uint64_t range_index) const = 0;
+
+  virtual const std::vector<Range>& get_ranges() const = 0;
+
+  /**
+   * Returns ``true`` if the current range is the default range.
+   *
+   * The default range is the full range for the dimension the range is
+   * managing.
+   **/
+  virtual bool is_default() const = 0;
+
+  /** Returns ``true`` if the range subset is the empty set. */
+  virtual bool is_empty() const = 0;
+
+  /**
+   * Returns ``false`` if the subset contains a range other than the default
+   * range.
+   **/
+  virtual bool is_set() const = 0;
+
+  /**
+   * Returns ``true`` if there is exactly one unary ranage in the subset.
+   */
+  virtual bool is_unary() const = 0;
+
+  /** Returns the number of distinct ranges stored in the range manager. */
+  virtual uint64_t num_ranges() const = 0;
+
+  /**
+   * Sorts the ranges in the range manager.
+   *
+   * @param compute_tp
+   */
+  virtual Status sort_ranges(ThreadPool* const compute_tp) = 0;
+};
+
+template <typename T, Datatype D, bool CoalesceAdds>
+class RangeSubset : public RangeSubsetBase {
+ private:
+  using AddStrategy = detail::AddStrategy<CoalesceAdds, T>;
+  using SortStrategy = detail::SortStrategy<D, T>;
+
+  /** Maximum possible range. */
+  Range full_range_;
+
+  /**
+   * If ``true``, the range contains the full domain for the dimension (the
+   * default value for a subarray before any other values is set. Otherwise,
+   * some values has been explicitly set to the range.
+   */
+  bool is_default_ = true;
+
+  /** Stored ranges. */
+  std::vector<Range> ranges_;
+
+ public:
+  /** Disable default constructor. */
+  RangeSubset() = delete;
+
+  /**
+   * Constructor for the RangeSubset.
+   *
+   * @param full_range The range this is a subset of.
+   * @param is_default Flag indicating if this is a default subset. The default
+   * subset is the full range.
+   */
+  RangeSubset(const Range& full_range, bool is_default)
+      : full_range_(full_range)
+      , is_default_(is_default)
+      , ranges_() {
+    if (is_default)
+      ranges_.emplace_back(full_range);
+  };
+
+  /** Destructor. */
+  ~RangeSubset() = default;
+
+  //  Status add_range(
+  //      Range&& range,
+  //      bool error_on_oob) {
+  //    if (!error_on_oob)
+  //      RETURN_NOT_OK(adjust_range_oob(&new_range));
+  //    RETURN_NOT_OK(check_range(&new_range));
+  //    add_range_unsafe(ranges, new_range);
+  //  };
+
+  Status add_range_unsafe(const Range& range) override {
+    if (is_default_) {
+      ranges_.clear();
+      is_default_ = false;
+    }
+    return AddStrategy::add_range(ranges_, range);
+  };
+
+  const Range& get_range(const uint64_t range_index) const override {
+    return ranges_[range_index];
+  };
+
+  const std::vector<Range>& get_ranges() const override {
```

Comment:
The convention for things like this is to omit `get_`. So either `ranges()` or maybe even `data()` like other containers use.